From 2a1e025e8d4d104966800e4d4b1652983231eaef Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Thu, 18 Sep 2014 14:20:53 -0400
Subject: [fs] nfsv3: Fix another acl regression

Message-id: <1411050057-12417-91-git-send-email-steved@redhat.com>
Patchwork-id: 96260
O-Subject: [RHEL7.1 PATCH 90/94] NFSv3: Fix another acl regression
Bugzilla: 1111708
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>

From: Trond Myklebust <trond.myklebust@primarydata.com>

When creating a new object on the NFS server, we should not be sending
posix setacl requests unless the preceding posix_acl_create returned a
non-trivial acl. Doing so, causes Solaris servers in particular to
return an EINVAL.

Fixes: 013cdf1088d72 (nfs: use generic posix ACL infrastructure,,,)
Resolves: https://bugzilla.redhat.com/show_bug.cgi?id=1132786
Cc: stable@vger.kernel.org # 3.14+
Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Upsteam Id: f87d928f6d98644d39809a013a22f981d39017cf
Bugzilla: 1111708
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfs/nfs3acl.c b/fs/nfs/nfs3acl.c
index 34e918d..7f0ed5e 100644
--- a/fs/nfs/nfs3acl.c
+++ b/fs/nfs/nfs3acl.c
@@ -305,7 +305,10 @@ static int nfs3_proc_setacls(struct inode *inode, struct posix_acl *acl,
   .rpc_argp = &args,
   .rpc_resp = &fattr,
  };
- int status;
+ int status = 0;
+
+ if (acl == NULL && (!S_ISDIR(inode->i_mode) || dfacl == NULL))
+  goto out;
 
  status = -EOPNOTSUPP;
  if (!nfs_server_capable(inode, NFS_CAP_ACLS))
-- 
1.7.1